llvm/lib/Object/XCOFFObjectFile.cpp

```diff
@@ -1242,15 +1242,72 @@ bool XCOFFSymbolRef::isFunction() const {
 
   const XCOFFCsectAuxRef CsectAuxRef = ExpCsectAuxEnt.get();
 
-  // A function definition should be a label definition.
-  // FIXME: This is not necessarily the case when -ffunction-sections is
-  // enabled.
-  if (!CsectAuxRef.isLabel())
+  if (CsectAuxRef.getStorageMappingClass() != XCOFF::XMC_PR &&
+      CsectAuxRef.getStorageMappingClass() != XCOFF::XMC_GL)
     return false;
 
-  if (CsectAuxRef.getStorageMappingClass() != XCOFF::XMC_PR)
+  // A function definition should not be a common type symbol or a external
+  // symbol.
+  if (CsectAuxRef.getSymbolType() == XCOFF::XTY_CM ||
+      CsectAuxRef.getSymbolType() == XCOFF::XTY_ER)
     return false;
 
+  // If the next symbol is a XTY_LD type symbol with same address, this XTY_SD
+  // symbol is not a function. Otherwise this is a function symbol for
+  // -ffunction-sections.
+  if (CsectAuxRef.getSymbolType() == XCOFF::XTY_SD) {
+    // If this is a csect with size 0, it won't be a function definition.
+    // This is used to hack situation that llvm always generates below symbol
+    // for -ffunction-sections:
+    // m   0x00000000     .text     1  unamex                    **No Symbol**
+    // a4  0x00000000       0    0     SD       PR    0    0
+    if (getSize() == 0)
+      return false;
+
+    uint8_t NumberOfAuxEntries = getNumberOfAuxEntries();
+
+    // If this is the last main symbol table entry, there won't be XTY_LD type
+    // symbol below.
+    if (getEntryAddress() == getObject()->getSymbolEntryAddressByIndex(
+                                 getObject()->getNumberOfSymbolTableEntries() -
+                                 NumberOfAuxEntries - 1))
+      return true;
+
+    DataRefImpl Ref;
+    Ref.p = XCOFFObjectFile::getAdvancedSymbolEntryAddress(
+        getEntryAddress(), NumberOfAuxEntries + 1);
+    XCOFFSymbolRef NextSym = getObject()->toSymbolRef(Ref);
+    if (!NextSym.isCsectSymbol())
+      return true;
+
+    Expected<uint64_t> SymbolAddressOrErr = getAddress();
+    if (!SymbolAddressOrErr)
+      return false;
+
+    Expected<uint64_t> NextSymbolAddressOrErr = NextSym.getAddress();
+    if (!NextSymbolAddressOrErr)
+      return true;
+
+    if (SymbolAddressOrErr.get() != NextSymbolAddressOrErr.get())
+      return true;
+
+    // Check next symbol is XTY_LD. If so, this symbol is not a function.
+    Expected<XCOFFCsectAuxRef> NextCsectAuxEnt = NextSym.getXCOFFCsectAuxRef();
+    if (!NextCsectAuxEnt) {
+      // If the next symbol has no aux entries, won't be a XTY_LD symbol.
+      consumeError(NextCsectAuxEnt.takeError());
```

Comment:
As a general rule, `consumeError` is a code smell. If the underlying code can't raise an error, then this should be `cantFail`. If it can raise an error, then that error should be reported (either by handling it here or passing it up the stack). `consumeError` should only be used if we really don't care that there's been an error.

Reply:
Thanks. Then I think we should use `consumeError`. There is a possibility that the csect symbol has no aux entries. And this is still a valid csect symbol but not a symbol that is searching for here.

Reply:
If it's valid to not have aux entries as you say then it seems to me like the code you're calling shouldn't be returning an Error in that case, but rather a result saying "there are none", as otherwise there's no way of distinguishing this from "there was a problem with the entries". Alternatively, there needs to be a way to count the number of entries before calling this function.

Reply:
Can we distinguish the "there are none" and "there was a problem with the entries" in `getXCOFFCsectAuxRef()` from different error string? I saw `getXCOFFCsectAuxRef()` creates different errors for them.

Checking number of the entries before calling `getXCOFFCsectAuxRef()` seems redundant because inside `getXCOFFCsectAuxRef()`, the number is also checked?

Reply:
According to the XCOFF documentation, a CSECT auxiliary symbol table entry is required for a C_EXT, C_HIDEXT, or C_WEAKEXT symbol.  Otherwise, the object file is not valid.

Reply:
thanks for clarify.

Reply:
So for someone not familiar with XCOFF does that mean that there are other symbol types that don't have an auxiliary entry? I.e. having no entries in this case is valid?

Reply:
For csect symbol, a valid symbol is required to have at least one aux entry.

However the function `getXCOFFCsectAuxRef()` also only check for `isCsectSymbol` but it reports several kinds of errors, for example a invalid object. I think here we should not assume there is no error reported from this function and should just skip these errors?

Reply:
> I think here we should not assume there is no error reported from this function

Seems reasonable.

> should just skip these errors

No, I don't think that's reasonable, unless there's a strong reason why the error shouldn't be reported up the stack. Otherwise, the client code (and ultimately the end user) will never realise there's a problem.

Reply:
Do you mean I should change the `isFunction()` to return `Expected<bool>`? I can do this if you think that makes more sense.

Reply:
That's what I was immediately thinking of, yes. I'd suggest changing it in a separate PR, so that we don't inflate the size of this one too much.

Reply:
https://github.com/llvm/llvm-project/pull/72232 is created for this comments. Once that nfc patch is committed, I will rebase this one. Thanks.

Reply:
Rebased and updated according to comments.